pkg/validation/validation.go

```diff
@@ -0,0 +1,220 @@
+		allErrs = append(allErrs, k8sValidation.ValidateNonnegativeField(int64(*spec.ActiveDeadlineSeconds), fieldPath.Child("activeDeadlineSeconds"))...)
+	}
+	if spec.Selector == nil {
+		allErrs = append(allErrs, k8sField.Required(fieldPath.Child("selector"), ""))
+	} else {
+		allErrs = append(allErrs, k8sValidationUnv.ValidateLabelSelector(spec.Selector, fieldPath.Child("selector"))...)
+	}
+	allErrs = append(allErrs, ValidateWorkflowSteps(spec.Steps, fieldPath.Child("steps"))...)
+	return allErrs
+}
+
+func topologicalSort(steps map[string]api.WorkflowStep, fieldPath *k8sField.Path) ([]string, *k8sField.Error) {
+	sorted := make([]string, len(steps))
+	temporary := map[string]bool{}
+	permanent := map[string]bool{}
+	cycle := []string{}
+	isCyclic := false
+	cycleStart := ""
+	var visit func(string) *k8sField.Error
+	visit = func(n string) *k8sField.Error {
+		if _, found := steps[n]; !found {
+			return k8sField.NotFound(fieldPath, n)
+		}
+		switch {
+		case temporary[n]:
+			isCyclic = true
+			cycleStart = n
+			return nil
+		case permanent[n]:
+			return nil
+		}
+		temporary[n] = true
+		for _, m := range steps[n].Dependencies {
+			if err := visit(m); err != nil {
+				return err
+			}
+			if isCyclic {
+				if len(cycleStart) != 0 {
+					cycle = append(cycle, n)
+					if n == cycleStart {
+						cycleStart = ""
+					}
+				}
+				return nil
+			}
+		}
+		delete(temporary, n)
+		permanent[n] = true
+		sorted = append(sorted, n)
+		return nil
+	}
+	var keys []string
+	for k := range steps {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		if permanent[k] {
+			continue
+		}
+		if err := visit(k); err != nil {
+			return nil, err
+		}
+		if isCyclic {
+			return nil, k8sField.Forbidden(fieldPath, fmt.Sprintf("detected cycle %s", cycle))
+		}
+	}
+	return sorted, nil
+}
+
+func ValidateWorkflowSteps(steps map[string]api.WorkflowStep, fieldPath *k8sField.Path) k8sField.ErrorList {
+	allErrs := k8sField.ErrorList{}
+	if _, err := topologicalSort(steps, fieldPath); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	for k, v := range steps {
+		if v.JobTemplate != nil && v.ExternalRef != nil {
+			allErrs = append(allErrs, k8sField.Invalid(fieldPath, k, "jobTemplate and externalRef are mutually exclusive"))
+		}
+	}
+	return allErrs
+}
+
+func ValidateWorkflowStatus(status *api.WorkflowStatus, fieldPath *k8sField.Path) k8sField.ErrorList {
+	allErrs := k8sField.ErrorList{}
+	return allErrs
+}
+
+func getWorkflowUnmodifiableSteps(workflow *api.Workflow) (running, completed map[string]bool) {
+	running = make(map[string]bool)
+	completed = make(map[string]bool)
+	if workflow.Status.Statuses == nil {
+		return
+	}
+	for key := range workflow.Spec.Steps {
+		if step, found := workflow.Status.Statuses[key]; found {
+			if step.Complete {
+				completed[key] = true
+			} else {
+				running[key] = true
+			}
+		}
+	}
+	return
+}
+
+func ValidateWorkflowUpdate(workflow, oldWorkflow *api.Workflow) k8sField.ErrorList {
```

Comment:
We should think about when to call this function. We will update the wf from our own sync loop. We probably only want to call this function when a user performs an update (or maybe not; lets discuss).